### PR TITLE
[SPARK-46258][CORE] Add `RocksDBPersistenceEngine`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to add `RocksDBPersistenceEngine`.

### Why are the changes needed?

To speed up `Spark Master` HA operations by **6.1x**.

```
OpenJDK 64-Bit Server VM 21.0.1+12-LTS on Linux 5.15.0-1051-azure
AMD EPYC 7763 64-Core Processor
1000 Workers:                                        Best Time(ms)   Avg Time(ms)   Relative
--------------------------------------------------------------------------------------------
FileSystemPersistenceEngine with JavaSerializer               1571           1616       3.6X
RocksDBPersistenceEngine with JavaSerializer                   257            258      22.0X
```

### Does this PR introduce _any_ user-facing change?

No. This is a new backend.

### How was this patch tested?

Pass the CIs with the newly added test cases.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR, @LuciferYang ?

**Comment by dongjoon-hyun** [comment]
Could you review this PR, @yaooqinn ?

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * @param serializer Used to serialize our objects.
+ */
+private[master] class RocksDBPersistenceEngine(
+    val dir: String,
+    val serializer: Serializer)
+  extends PersistenceEngine with Logging {
+
+  RocksDB.loadLibrary()
+
+  val path = Files.createDirectories(Paths.get(dir))
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:46]: due to this a `public` member

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala [hunk 2]
```diff
@@ -67,6 +67,25 @@ private[master] class FileSystemRecoveryModeFactory(conf: SparkConf, serializer:
...
 }
 
+/**
+ * LeaderAgent in this case is a no-op. Since leader is forever leader as the actual
+ * recovery is made by restoring from RocksDB.
+ */
+private[master] class RocksDBRecoveryModeFactory(conf: SparkConf, serializer: Serializer)
+  extends StandaloneRecoveryModeFactory(conf, serializer) with Logging {
+
+  val recoveryDir = conf.get(RECOVERY_DIRECTORY)
```
Comment (yaooqinn) [inline core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala:77]: It looks better to have a separate directory from the FileSystemRecoveryModeFactory

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  private val options = new Options()
+    .setCreateIfMissing(true)
+    .setBottommostCompressionType(CompressionType.ZSTD_COMPRESSION)
+    .setCompressionType(CompressionType.LZ4_COMPRESSION)
+    .setTableFormatConfig(tableFormatConfig)
+
+  private val db: RocksDB = RocksDB.open(options, path.toString)
+
+  override def persist(name: String, obj: Object): Unit = {
+    val serialized = serializer.newInstance().serialize(obj)
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:75]: Can `db.put(name.getBytes(UTF_8), serialized.array())` be executed directly when ` serialized.hasArray` is true? Is this a feasible optimization?

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  override def read[T: ClassTag](name: String): Seq[T] = {
+    val result = new ArrayBuffer[T]
+    val iter = db.newIterator()
+    iter.seek(name.getBytes(UTF_8))
+    while (iter.isValid && new String(iter.key()).startsWith(name)) {
+      result.append(serializer.newInstance().deserialize[T](ByteBuffer.wrap(iter.value())))
+      iter.next()
+    }
+    iter.close()
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:93]: should we close iter in finally block?

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala:77]: +1

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 5]
```diff
@@ -186,6 +186,10 @@ private[deploy] class Master(
         val fsFactory =
           new FileSystemRecoveryModeFactory(conf, serializer)
         (fsFactory.createPersistenceEngine(), fsFactory.createLeaderElectionAgent(this))
+      case "ROCKSDB" =>
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala:189]: Is this configuration item required to be `ROCKSDB`? Is it not allowed to configure it as `rocksdb` or `Rocksdb` now? （Perhaps this is a legacy issue?）

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala:77]: Thank you for review. @yaooqinn and @LuciferYang 

I tried according to your advices but it seems to make the configuration namespace a little weird because this existing configuration is a general one, `spark.deploy.recoveryDirectory`.

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/core/src/main/scala/org/apache/spark/internal/config/Deploy.scala#L54

If we introduce a new configuration for `RocksDBPersistenceEngine` like the following, the users will ask why we cannot use `spark.deploy.recoveryDirectory` for `ROCKSDB` mode. Here is the example. The AS-IS design aims to allow the users to switch a single configuration `recoveryMode`.

**BEFORE**
```
spark.deploy.recoveryMode=FILESYSTEM
spark.deploy.recoveryDirectory=/opt/master
```

**AFTER**
```
spark.deploy.recoveryMode=ROCKSDB
spark.deploy.recoveryDirectory=/opt/master
```

If we add a new configuration, it would be the following where `spark.deploy.recoveryDirectory` is no-op.
```
spark.deploy.recoveryMode=ROCKSDB
spark.deploy.recoveryDirectory=/opt/master
spark.deploy.recoveryRocksDBDirectory=/opt/db
```

In addition, we assume that the users clear the location properly when they changes the setting like the following. 
- `spark.deploy.recoverySerializer`
- `spark.deploy.recoveryCompressionCodec`

To sum up, if you guys don't mind, I'd like to use the existing general configuration name for this backend.

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:46]: Thanks. Let me switch it to `private`.

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:93]: Sure.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:189]: Yes, it's a legacy behavior which `spark.deploy.recoveryMode` always assumes all upper cases from Spark 0.8.1.
- NONE, ZOOKEEPER, FILESYSTEM, CUSTOM

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:75]: This one, right?
```scala
  override def persist(name: String, obj: Object): Unit = {
    val serialized = serializer.newInstance().serialize(obj)
    if (serialized.hasArray) {
      db.put(name.getBytes(UTF_8), serialized.array())
    } else {
      val bytes = new Array[Byte](serialized.remaining())
      serialized.get(bytes)
      db.put(name.getBytes(UTF_8), bytes)
    }
  }
```

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala:77]: Sure.

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 3 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:75]: Yes

**Comment by LuciferYang** [comment]
Maybe the documents also need to be updated in this pr?

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/docs/spark-standalone.md?plain=1#L736

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala]: No, it will be okay. For example, rocksdb content will be ignored because FILESYSTEM scans only the following `prefix` files. I guess RocksDB also do the same things for the FILESYSTEM's leftover.

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/core/src/main/scala/org/apache/spark/deploy/master/PersistenceEngine.scala#L56-L78

However, that kind of changes are not supposed to be there. For the following three configurations, the directory should be cleared.

- spark.deploy.recoveryMode
- spark.deploy.recoverySerializer
- spark.deploy.recoveryCompressionCodec

**Comment by dongjoon-hyun** [comment]
Thank you for pointing out that.
> Maybe the documents also need to be updated in this pr?

Actually, I thought about making a new doc PR for the following accumulated stuffs. I also want to add some recommendation for the user choices. Let me proceed them separately~
```
[SPARK-46258][CORE] Add RocksDBPersistenceEngine
[SPARK-46216][CORE] Improve `FileSystemPersistenceEngine` to support compressions
[SPARK-46205][CORE] Improve `PersistenceEngine` performance with `KryoSerializer`
```

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala]: The thing is, it's pretty common to see that old naming conventions just don't cut it when adding new features.

Technically, we have the `spark.deploy.zookeeper.dir` for ZK-mode although it's a znode. 

I don't have a strong opinion for a new configuration here as it's trivial.

**Comment by dongjoon-hyun** [comment]
Thank you, @LuciferYang .

File: core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RecoveryModeFactory.scala]: Thank you, @yaooqinn . Ya, that's true. ZookeeperPersistenceEngine has a separate configuration namespace for that. That's one of the reasons why I don't want to add a new namespace.

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    .setBottommostCompressionType(CompressionType.ZSTD_COMPRESSION)
+    .setCompressionType(CompressionType.LZ4_COMPRESSION)
+    .setTableFormatConfig(tableFormatConfig)
+
+  private val db: RocksDB = RocksDB.open(options, path.toString)
+
+  override def persist(name: String, obj: Object): Unit = {
+    val serialized = serializer.newInstance().serialize(obj)
+    if (serialized.hasArray) {
+      db.put(name.getBytes(UTF_8), serialized.array())
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:77]: When do you think `persist` is invoked with the duplication?

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:77]: To @yaooqinn , `PersistenceEngine` is a secondary storage instead of lookup tables. In general, `Master` has a way to keep the uniqueness in the memory structure before reaching here.

For example, `WorkerInfo` is under tracking of 

https://github.com/apache/spark/blob/1d80d80a56c418f841e282ad753fad6671c3baae/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L76C9-L76C9
```
private val idToWorker = new HashMap[String, WorkerInfo]
```

In addition, Driver IDs and App IDs are generated by Master in the unique way including timestamps.

**Comment by dongjoon-hyun** [comment]
I updated the benchmark result with the final commit, too~

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:75]: To @LuciferYang , the benchmark is updated with your suggestion.

File: core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/deploy/master/RocksDBPersistenceEngine.scala:77]: > When do you think persist is invoked with the duplication?

I guess it won't happen. I'm just asking because of the existing test `SPARK-46191: FileSystemPersistenceEngine.persist error message for the existing file`

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you so much both of you, @LuciferYang and @yaooqinn . Merged to master.
